Add turborepo

barbarah opened the pull request:
- [x] add Turborepo
- [x] Move the dataset browser to the app workspace
- [x] Use the three default turbo packages, and update with our config
   - [x] eslint-config-custom
   - [x] tsconfig
   - [x] UI
      - [x]  Add the components, now placed under `/components`

TODO

- [ ] fix the vercel build
- [x] Check the GitHub workflows
- [x] update readme
- [x] What to do with the docs folder?
- [x] Docker

barbarah committed:
Move the dataset browser to new apps directory and init Turborepo

barbarah committed:
Fix typescript path

barbarah committed:
Move prettierrc.js to root directory

barbarah committed:
Only keep root package-lock.json

barbarah committed:
Use tsconfig workspace package

barbarah committed:
configure eslint with turborepo

barbarah committed:
Add tailwind to ui package

barbarah committed:
fix the versions of all packages

barbarah committed:
Move global components to ui package

barbarah committed:
Remove the example apps

barbarah committed:
Move tailwind config to separate package

barbarah committed:
Move cypress tsconfig

barbarah committed:
Set the node version correct

barbarah committed:
Set eslint rules for ui package

barbarah committed:
Re-add unintentional removed tailwind config

sdevalk commented on apps/dataset-browser/.gitignore:
```diff
@@ -0,0 +1,36 @@
+# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
```
Do we need this `.gitignore`? Or can we use the one in the root?

barbarah replied:
Lets use just one

sdevalk commented on apps/dataset-browser/.vscode/settings.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Curious: can we put this file in the root of the repo (so that it can be used by all apps)?

barbarah replied:
This file is automatically created after `npm run dev`. This is not customizable, and there is already an issue here: https://github.com/vercel/next.js/issues/42213 

For now, I can do nothing, but I will subscribe to the issue to know if there are any updates.

sdevalk commented on apps/dataset-browser/tsconfig.json:
```diff
@@ -0,0 +1,11 @@
+      "@/*": ["./src/*"]
+    }
+  },
+  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
```
Curious: `next-env.d.ts` is already set in the parent tsconfig. Is it automatically used or must we explicitly set it again here?

barbarah replied:
We can remove this

sdevalk commented on package.json:
```diff
@@ -1,59 +1,25 @@
+    "build": "turbo run build",
+    "dev": "turbo run dev",
+    "lint": "turbo run lint",
+    "format": "prettier --write \"**/*.{ts,tsx,md}\""
```
Make use of gts' `fix` command here instead?

sdevalk commented on package.json:
```diff
@@ -1,59 +1,25 @@
-    "tailwindcss": "3.2.7",
-    "typescript": "5.0.3"
-  }
+    "prettier": "2.8.7",
```
I think `gts` already brings prettier. Do we need to add it here?

sdevalk commented on packages/tsconfig/base.json:
```diff
@@ -0,0 +1,16 @@
+  "display": "Default",
+  "extends": "../../node_modules/gts/tsconfig-google.json",
+  "compilerOptions": {
+    "composite": false,
```
Is this the default? If so, remove it?

sdevalk commented on packages/ui/package.json:
```diff
@@ -0,0 +1,27 @@
+  "types": "./index.tsx",
+  "license": "MIT",
+  "scripts": {
+    "lint": "eslint \"**/*.ts*\""
```
Is it possible to make use of `gts` here?

sdevalk commented on packages/ui/package.json:
```diff
@@ -0,0 +1,27 @@
+  "devDependencies": {
+    "@types/react": "18.0.33",
+    "autoprefixer": "10.4.14",
+    "eslint": "8.37.0",
```
I think `gts` already brings eslint. Do we need to add it explicitly here?

barbarah committed:
Only use one .gitignore

barbarah commented:
I noticed that `gts fix` and `gts link` wasn't working in the npm workspaces. This is related to the issues:

- https://github.com/google/gts/issues/752
- https://github.com/google/gts/issues/490

The last issue is with `yarn` but has more discussion. One solution was using `nohoist`. This is not a default option with npm. So I tried with the following package https://github.com/zgriesinger/noist. But this did not fix the errors. The last solution mentioned was only linting from the root level; this works.

I have also subscribed to the issue [[feature] support npm workspaces](https://github.com/google/gts/issues/718). There may be a better solution in the future.

barbarah committed:
Make gts work with workspaces

I noticed that gts fix and gts link wasn't working in the npm workspaces. This is related to the issues:

gts monorepo does not work with npm ? google/gts#752
gts does not work in a yarn workspaces monorepo google/gts#490
The last issue is with yarn but has more discussion. One solution was using nohoist. This is not a default option with npm. So I tried with the following package https://github.com/zgriesinger/noist. But this did not fix the errors. The last solution mentioned was only linting from the root level; this works.

I have also subscribed to the issue [feature] support npm workspaces. There may be a better solution in the future.

barbarah committed:
Remove package prettier

barbarah committed:
Harmonize package.json

barbarah commented on packages/tailwind-config/package.json:
```diff
@@ -0,0 +1,10 @@
+  "name": "tailwind-config",
+  "version": "0.0.0",
+  "private": true,
+  "main": "index.js",
```
No, I will remove this line

barbarah committed:
Remove non existing main script form package.json

barbarah committed:
Update tsconfig package

barbarah committed:
Remove eslint from ui package

barbarah committed:
Move the docs to the root level

barbarah committed:
Update readme

barbarah committed:
fix github workflows

barbarah commented:
I have updated the workflows or added the needed scripts to package.json. I cannot see if the workflows are working. The Vercel build first needs to be successful.

barbarah committed:
Merge remote-tracking branch 'origin/main' into 191-add-turborepo

# Conflicts:
#       package-lock.json
#       package.json

barbarah committed:
Use versioning-strategy increase for dependabot

sdevalk commented on .github/workflows/end-to-end-testing.yml:
Or should we make an app-specific file for this, so as to make its scope clear (e.g. `end-to-end-testing-dataset-browser.yml`)?

sdevalk commented on README.md:
```diff
@@ -1,4 +1,17 @@
+
+## What's inside?
+
+This repo uses [turborepo](https://turbo.build/) as build system and [npm](https://www.npmjs.com/) as a package manager. It includes the following packages/apps:
```
Minor: `...following apps and packages`

sdevalk commented on README.md:
```diff
@@ -1,4 +1,17 @@
+- `ui`: a React component library shared by the apps
+- `eslint-config-custom`: `eslint` configurations
+- `tsconfig`: `tsconfig.json`s used throughout the monorepo
+- `tailwind-config`: tailwind config used by both the apps and the ui package
```
Minor: `ui` package (otherwise: UI)

sdevalk commented on README.md:
```diff
@@ -7,19 +20,25 @@
 
 ## Without Docker
 
+The following commands will run for all the workspaces. If you want to run a command for one workspace add the -w command. For example, to add a package to the dataset-browser:
```
Minor: add two newlines after the first sentence?

sdevalk commented on README.md:
```diff
@@ -7,19 +20,25 @@
 ### Install packages
 
-    npm install --no-progress
+Install all the packages, including the apps and packages
```
This sentence contains the word 'packages' two times - reformulate a little bit? Or just remove ', including the apps and packages'?

sdevalk commented on README.md:
```diff
@@ -7,19 +20,25 @@
     npm run dev
 
-Open [http://localhost:3000](http://localhost:3000) with your browser to see the result.
+Opens the dataset-browser on [http://localhost:3000](http://localhost:3000).
```
'Dataset Browser'?

sdevalk commented on docs/components.md:
```diff
@@ -1,14 +1,14 @@
 ## Generic components
 
-Components are generic if they are being used on multiple pages. Place these components in the `/components` folder.
+Components are generic if they are being used on multiple apps. Place these components in the UI package at `/packages/ui`.
```
'used in multiple apps'?

sdevalk commented on docs/components.md:
```diff
@@ -1,14 +1,14 @@
+## App-specific components
 
-Place the page-specific components in the `/app` folder at the same level as the page that uses the component.
+Place the page-specific components in the `/apps/[app-name]/src/app` folder at the same level as the page that uses the component.
```
Or: 'place the app-specific components...'?

sdevalk commented on packages/tsconfig/nextjs.json:
```diff
@@ -0,0 +1,24 @@
+    ],
+    "allowJs": true,
+    "skipLibCheck": true,
+    "strict": true,
```
`gts/tsconfig-google.json` already sets `strict: true` - can we remove it here?

sdevalk commented on packages/tsconfig/nextjs.json:
```diff
@@ -0,0 +1,24 @@
+      }
+    ],
+    "allowJs": true,
+    "skipLibCheck": true,
```
`base.json` already sets this setting - can we remove it here?

sdevalk commented on packages/tsconfig/nextjs.json:
```diff
@@ -0,0 +1,24 @@
+    "strict": true,
+    "noEmit": true,
+    "incremental": true,
+    "esModuleInterop": true,
```
`base.json` already sets this setting - can we remove it here?

sdevalk commented on packages/tsconfig/react-library.json:
```diff
@@ -0,0 +1,9 @@
+  "extends": "./base.json",
+  "compilerOptions": {
+    "jsx": "react-jsx",
+    "module": "ESNext"
```
`esnext` (for consistency with `nextjs.json`)?

sdevalk commented on packages/eslint-config-custom/package.json:
```diff
@@ -0,0 +1,17 @@
+  "main": "index.js",
+  "private": true,
+  "devDependencies": {
+    "eslint": "8.37.0",
+    "eslint-config-next": "13.3.0",
+    "eslint-config-prettier": "8.8.0",
```
`gts` already brings `eslint` and `eslint-config-prettier` - do we need to install these here?

barbarah committed:
Typos and grammatical changes

barbarah committed:
Removed packages already in root level

barbarah committed:
Update tsconfig

barbarah commented:
This work is moved to https://github.com/colonial-heritage/colonial-collections
